### ISSUE TITLE
DB Connection Closes Unexpectedly

### ISSUE THREAD

**Issue by amihaiemil** [self-xdsd/self-todos#63]
Sometimes the DB Connection is lost, let's fix it:

```java
2021-05-21 12:22:36.931  WARN 9875 --- [http-nio-8282-exec-5] com.selfxdsd.todos.SelfCoreComponent     : Closing Self Core...
2021-05-21 12:22:36.933  WARN 9875 --- [http-nio-8282-exec-5] com.selfxdsd.todos.SelfCoreComponent     : Self Core successfully closed!
2021-05-21 12:22:36.940 ERROR 9875 --- [task-3] .a.i.SimpleAsyncUncaughtExceptionHandler : Unexpected exception occurred invoking async method: public void com.selfxdsd.todos.PuzzlesComponent.review(com.selfxdsd.api.Event)

org.jooq.exception.DataAccessException: SQL [select `self_xdsd`.`slf_jsonstorage_xdsd`.`url`, `self_xdsd`.`slf_jsonstorage_xdsd`.`etag`, `self_xdsd`.`slf_jsonstorage_xdsd`.`jsonBody` from `self_xdsd`.`slf_jsonstorage_xdsd` where `self_xdsd`.`slf_jsonstorage_xdsd`.`url` = ?]; No operations allowed after connection closed.
	at org.jooq_3.13.2.MYSQL.debug(Unknown Source) ~[na:na]
	at org.jooq.impl.Tools.translate(Tools.java:2751) ~[jooq-3.13.2.jar!/:na]

```

**Comment by zoeself** [comment]
@amihaiemil thank you for reporting this. Unfortunately, @amihaiemil is not a contributor with the required role (DEV), that's why I unassigned them. 

I will assign a contributor to take care of it soon.

**Comment by amihaiemil** [comment]
The problem is that the puzzles are reviewed in an Async method and the RequestScoped DB connection closes before the jsonstorage is accessed (for e-tag caching).

**Comment by amihaiemil** [comment]
@criske any idea how to fix this? :D

**Comment by zoeself** [comment]
@amihaiemil I couldn't find any assignee for this task. This is either because there are no contributors with role ``DEV`` available or because the project does not have enough funds.

Please, make sure there is at least one available contributor with the required role and the project can afford to pay them.

**Comment by criske** [comment]
@amihaiemil maybe extract the Database component in a prototype scope?

**Comment by amihaiemil** [comment]
@criske I have no idea what ``prototype`` scope is hihi. I was thinking: make the SelfCoreComponent SessionScoped and pass it as argument to the Async method, which will call close() on it manually at the end

**Comment by criske** [comment]
@amihaiemil It might work, but looks dirty :D

but I  think making SelfCoreComponent prototype will be suffice (this mean one connection per injection)
this means that it will be not shared by PuzzlesApi controller and PuzzlesComponent.

Since SelfCompoent is autclosable I think will be treated as DisposableBean by Spring and thus not need to manually call close.


i need to test though.. to be sure :D

**Comment by amihaiemil** [comment]
@criske sounds like a better plan :D

**Comment by zoeself** [comment]
@criske this is your task now, please go ahead. Deadline (when this ticket should be closed) is ``2021-05-31T13:36:40.770107``.

Estimation here is ``60 minutes``, that's how much you will be paid.

Remember, you don't have to solve everything in this ticket - you can solve it [partially](https://docs.self-xdsd.com/microtasking.html) and leave ``todo`` markers in the code, which will become future tasks.

If you have any questions don't ask me, I'm not a technical person. Open new tickets instead.

If you don't have time or simply don't want to work on this, you can always [resign](https://docs.self-xdsd.com/projectmanager.html#commands).

**Comment by zoeself** [comment]
@criske thank you for resolving this ticket. I've just added it to your active invoice. You can always check all your invoices and more on the [Contributor Dashboard](https://self-xdsd.com).
